storm-core/src/jvm/backtype/storm/spout/CheckpointSpout.java

```diff
@@ -0,0 +1,280 @@
+    public static final String CHECKPOINT_ACTION_PREPARE = "prepare";
+    public static final String CHECKPOINT_ACTION_COMMIT = "commit";
+    public static final String CHECKPOINT_ACTION_ROLLBACK = "rollback";
+    public static final String CHECKPOINT_ACTION_INITSTATE = "initstate";
+
+    private static final String TX_STATE_KEY = "__state";
+    private static final int DEFAULT_CHECKPOINT_INTERVAL = 1000; // every sec
+
+    private TopologyContext context;
+    private SpoutOutputCollector collector;
+    private long lastCheckpointTs;
+    private int checkpointInterval;
+    private boolean recoveryStepInProgress;
+    private boolean checkpointStepInProgress;
+    private boolean recovering;
+    private KeyValueState<String, CheckPointState> checkpointState;
+
+    @Override
+    public void open(Map conf, TopologyContext context, SpoutOutputCollector collector) {
+        open(context, collector, loadCheckpointInterval(conf), loadCheckpointState(conf, context));
+    }
+
+    // package access for unit test
+    void open(TopologyContext context, SpoutOutputCollector collector,
+              int checkpointInterval, KeyValueState<String, CheckPointState> checkpointState) {
+        this.context = context;
+        this.collector = collector;
+        this.checkpointInterval = checkpointInterval;
+        this.checkpointState = checkpointState;
+        lastCheckpointTs = 0;
+        recoveryStepInProgress = false;
+        checkpointStepInProgress = false;
+        recovering = true;
+    }
+
+    @Override
+    public void nextTuple() {
+        if (shouldRecover()) {
+            LOG.debug("In recovery");
+            handleRecovery();
+            startProgress();
+        } else if (shouldCheckpoint()) {
+            LOG.debug("In checkpoint");
+            doCheckpoint();
+            startProgress();
+        }
+    }
+
+    @Override
+    public void ack(Object msgId) {
+        CheckPointState txState = getTxState();
+        LOG.debug("Got ack with txid {}, current txState {}", msgId, txState);
+        if (txState.txid == ((Number) msgId).longValue()) {
+            if (recovering) {
+                handleRecoveryAck();
+            } else {
+                handleCheckpointAck();
+            }
+        } else {
+            LOG.warn("Ack msgid {}, txState.txid {} mismatch", msgId, txState.txid);
+        }
+        resetProgress();
+    }
+
+    @Override
+    public void fail(Object msgId) {
+        LOG.debug("Got fail with msgid {}", msgId);
+        resetProgress();
+    }
+
+    @Override
+    public void declareOutputFields(OutputFieldsDeclarer declarer) {
+        declarer.declareStream(CHECKPOINT_STREAM_ID, new Fields(CHECKPOINT_FIELD_TXID, CHECKPOINT_FIELD_ACTION));
+    }
+
+    @Override
+    public Map<String, Object> getComponentConfiguration() {
+        Config conf = new Config();
+        conf.put(Config.TOPOLOGY_SLEEP_SPOUT_WAIT_STRATEGY_TIME_MS, 100);
```

Comment:
It was added to sleep for slightly longer time rather than spinning in the nextTuple. Instead of hard coding, will change this to 10% of the checkpoint interval time.